### PR TITLE
Add brotli to skipper

### PR DESCRIPTION
**Pull request by fpozzobon**
Adding brotli compression / decompression to skipper

References: https://github.com/zalando/skipper/issues/696

### CODE REVIEW

File: filters/builtin/compress.go [hunk 1]
```diff
@@ -268,9 +272,20 @@ func unsupported() {
 
 func newEncoder(enc string, level int) (encoder, error) {
 	switch enc {
+	case "br":
+		if level > brotli.BestCompression {
```
Comment (szuecs) [inline filters/builtin/compress.go:276]: this check you already did on filter creation (CreateFilter())

Comment (fpozzobon) [reply filters/builtin/compress.go]: removed :)

File: filters/builtin/compress.go [hunk 2]
```diff
@@ -366,11 +390,20 @@ func (c *compress) Response(ctx filters.FilterContext) {
...
 	if enc == "" {
 		return
 	}
 
 	responseHeader(rsp, enc)
 	responseBody(rsp, enc, c.level)
 }
+
+func index(array []string, value string) int {
+	for k, v := range array {
```
Comment (szuecs) [inline filters/builtin/compress.go:403]: nit: k->i 
i is index
k is key

File: filters/builtin/compress.go [hunk 3]
```diff
@@ -219,11 +232,12 @@ func acceptedEncoding(r *http.Request) string {
 		}
 
 		name := strings.ToLower(strings.TrimSpace(sp[0]))
-		if !stringsContain(supportedEncodings, name) {
+		prio := index(c.encoding, name)
```
[the remark this answers is not in the document]
Comment (AlexanderYastrebov) [reply filters/builtin/compress.go:235]: If we are looking to have a priority for the same quality value (or absence thereof) then I think we should look at stable sort https://pkg.go.dev/sort#Stable that preserves client preferred algorithm order from the header.

[the remark this answers is not in the document]
Comment (szuecs) [reply filters/builtin/compress.go:235]: @AlexanderYastrebov clients tell they can do X,Y,Z and the priority gives the owner of skipper a choice what algorithm should be preferred. For example you might want to use brotli as first choice and fallback to gzip. A Browser sends in some random order what they support, and then you get something you are lucky and have no control.

[the remark this answers is not in the document]
Comment (AlexanderYastrebov) [reply filters/builtin/compress.go:235]: > A Browser sends in some random order what they support, and then you get something you are lucky and have no control.

Looking at https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Accept-Encoding I think client sends its preferences using quality value for ordering (default quality value when absent is 1). The server serves the client so I think the logic should obey client preferences, i.e. use provided or default quality value without any additional server-side priority.

[the remark this answers is not in the document]
Comment (fpozzobon) [reply filters/builtin/compress.go:235]: @AlexanderYastrebov you are correct, the priority is kept on `quality value` if present but
checking on my Chrome browser, I can see the following:
`Accept-Encoding: gzip, deflate, br`

=> does it mean no preference or that `gzip` should be prioritised over `br`?

[the remark this answers is not in the document]
Comment (AlexanderYastrebov) [reply filters/builtin/compress.go:235]: I would say so, but I don't see order-dependent preference specified anywhere.
This answer https://stackoverflow.com/questions/3225136/http-what-is-the-preferred-accept-encoding-for-gzip-deflate also says first should be preferred but provides no reference.

If we would like to have a server side preference list then it would be interesting to address 
- why would brotli have priority over gzip (given that chrome does not do that and that it would be a change of behavior for the existing configurations)
- should the server-side priority list be configurable

E.g. Caddy implements configurable server side preference afaict from https://github.com/caddyserver/caddy/issues/3692

[the remark this answers is not in the document]
Comment (aryszka) [reply filters/builtin/compress.go:235]: I prefer the approach without the additional p field, and using stable sort to preserve the client specified order (which would be a fix of the current state in master)

[the remark this answers is not in the document]
Comment (AlexanderYastrebov) [reply filters/builtin/compress.go:237]: We've discussed this with @aryszka and came to conclusion that we should have configurable server-side priority for the algorithms in case quality values are equal (including default 1).

The default priority ordering should put brotly at the end (i.e. the priority should be `gzip > deflate > br`) to avoid unexpected algorithm change for the existing deployments.
Skipper operators then would be able to change the priority using a flag, something like `-compress-encoding-prefer=br,gzip,deflate` or e.g. exclude an algorithm like `-compress-encoding-prefer=br,gzip`.

File: skipper.go [hunk 4]
```diff
@@ -1456,6 +1459,10 @@ func run(o Options, sig chan os.Signal, idleConnsCH chan struct{}) error {
 		)
 	}
 
+	if len(o.CompressEncodings) > 0 {
+		o.CustomFilters = append(o.CustomFilters, builtin.NewCompressWithEncodings(o.CompressEncodings))
```
Comment (AlexanderYastrebov) [inline skipper.go:1463]: I think we need to remove `NewCompress()` from built-in registry to avoid name clash

File: filters/builtin/compress.go [hunk 5]
```diff
@@ -20,14 +21,16 @@ const bufferSize = 8192
...
 }
 
 type encodings []*encoding
 
 type compress struct {
-	mime  []string
-	level int
+	mime     []string
+	level    int
+	encoding []string
```
Comment (AlexanderYastrebov) [inline filters/builtin/compress.go:33]: I suggest to use a map here like
```go
map[string]int {
  "br": 1,
  "gzip": 2,
}
```
this should simplify priority lookup and would not require auxiliary `index` method

File: filters/builtin/compress.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (fpozzobon) [reply filters/builtin/compress.go:237]: thanks a lot for your feedbacks :)
I would suggest `compress-encodings` in place of `compress-encoding-prefer` (shorter)

File: config/config.go [hunk 6]
```diff
@@ -318,6 +320,7 @@ func NewConfig() *Config {
 	flag.Var(cfg.PredicatePlugins, "predicate-plugin", "set a custom predicate plugins to load, a comma separated list of name and arguments")
 	flag.Var(cfg.DataclientPlugins, "dataclient-plugin", "set a custom dataclient plugins to load, a comma separated list of name and arguments")
 	flag.Var(cfg.MultiPlugins, "multi-plugin", "set a custom multitype plugins to load, a comma separated list of name and arguments")
+	flag.Var(cfg.CompressEncodings, "compress-encodings", "set encodings supported for compression, when client has no preference, uses in priority in the order of this array")
```
Comment (fpozzobon) [inline config/config.go:323]: @AlexanderYastrebov is the description clear on the expected behaviour?

File: filters/builtin/compress.go [hunk 7]
```diff
@@ -123,28 +138,28 @@ func (e encodings) Swap(i, j int)      { e[i], e[j] = e[j], e[i] }
 //
 // The compression happens in a streaming way, using only a small internal buffer.
 //
-func NewCompress() filters.Spec { return &compress{} }
+func NewCompress() filters.Spec { return &compress{encoding: supportedEncodings} }
```
Comment (AlexanderYastrebov) [inline filters/builtin/compress.go:141]: Lets call `NewCompressWithEncodings()` here

File: filters/builtin/compress.go [hunk 8]
```diff
@@ -123,28 +138,28 @@ func (e encodings) Swap(i, j int)      { e[i], e[j] = e[j], e[i] }
...
+
+func NewCompressWithEncodings(encoding []string) filters.Spec { return &compress{encoding: encoding} }
```
Comment (AlexanderYastrebov) [inline filters/builtin/compress.go:143]: Would be nice to check that argument contains only supported values

File: skipper.go (hunk 4 above)
Comment (AlexanderYastrebov) [reply skipper.go:1463]: I think we should not remove it from `builtin.MakeRegistry()` as it could potentially break library users (but I think we've done this for some other filters in the past that required new config), @szuecs @aryszka WDYT?

Maybe its fine to put it into custom filters, we would need to update a comment below (and/or add a log to `registry.Register`) to hint that `CustomFilters` may override built-in.

File: config/config.go (hunk 6 above)
Comment (AlexanderYastrebov) [reply config/config.go:323]: Maybe something like
`set encodings supported for compression, the order defines priority when Accept-Header has equal quality values, see RFC 7231 section 5.3.1`

File: config/config.go [hunk 9]
```diff
@@ -281,6 +282,7 @@ func NewConfig() *Config {
 	cfg.RoutesURLs = commaListFlag()
 	cfg.ForwardedHeadersList = commaListFlag()
 	cfg.ForwardedHeadersExcludeCIDRList = commaListFlag()
+	cfg.CompressEncodings = commaListFlag("gzip", "br", "deflate")
```
Comment (AlexanderYastrebov) [inline config/config.go:285]: I think the default should be `"gzip", "deflate", "br"` to keep backward compatibility

File: docs/reference/filters.md [hunk 10]
```diff
@@ -365,14 +365,14 @@ to be omitted:
...
+no-compression, 1 means best-speed and 11 means best-compression. Example:
 
 ```
-* -> compress(9, "image/tiff") -> "https://www.example.org"
+* -> compress(11, "image/tiff") -> "https://www.example.org"
 ```
 
 The filter also checks the incoming request, if it accepts the supported encodings,
-explicitly stated in the Accept-Encoding header. The filter currently supports `gzip`
+explicitly stated in the Accept-Encoding header. The filter currently supports `gzip`, `brotli`
```
Comment (AlexanderYastrebov) [inline docs/reference/filters.md:375]: We should mention config option here

File: docs/kubernetes/ingress-usage.md [hunk 11]
```diff
@@ -424,13 +424,13 @@ Add a HTTP header in the response path of your clients.
 
     setResponseHeader("X-Foo", "bar")
 
-### Enable gzip
+### Enable compression
 
-Compress responses with gzip.
+Compress responses with gzip/brotli (depending on accepted encoding).
```
Comment (AlexanderYastrebov) [inline docs/kubernetes/ingress-usage.md:429]: Maybe we should just link to the filter docs here to not repeat the details

File: filters/builtin/compress.go (hunk 5 above)
Comment (szuecs) [reply filters/builtin/compress.go:33]: no that's more complex than having a slice that has an order already

File: skipper.go (hunk 4 above)
Comment (szuecs) [reply skipper.go:1463]: We could also move NewCompress() to the else branch of the code here and this should not break skipper library users.

File: config/config.go (hunk 9 above)
Comment (fpozzobon) [reply config/config.go:285]: sure I can change, it's only to enforce the allowed options here

File: filters/builtin/compress.go [hunk 12]
```diff
@@ -104,47 +119,60 @@ func (e encodings) Swap(i, j int)      { e[i], e[j] = e[j], e[i] }
...
-// encoding and sets the Vary: Accept-Encoding header, if missing.
+// with chunked transfer encodingPriority, sets the Content-Encoding to the selected
+// encodingPriority and sets the Vary: Accept-Encoding header, if missing.
 //
 // The compression happens in a streaming way, using only a small internal buffer.
 //
-func NewCompress() filters.Spec { return &compress{} }
+func NewCompress() filters.Spec { return NewCompressWithEncodings(supportedEncodings) }
+
+func NewCompressWithEncodings(encoding []string) filters.Spec {
```
Comment (AlexanderYastrebov) [inline filters/builtin/compress.go:147]: Since we are introducing options, lets make it future proof by using `Options` struct and `NewCompressWithOptions`, otherwise we would need to invent a new constructor for each new option

File: filters/builtin/compress.go [hunk 13]
```diff
@@ -37,8 +40,8 @@ type encoder interface {
 }
 
 var (
-	supportedEncodings  = []string{"gzip", "deflate"}
-	unsupportedEncoding = errors.New("unsupported encoding")
+	supportedEncodings  = []string{"gzip", "deflate", "br"}
+	unsupportedEncoding = errors.New("unsupported encodingPriority")
```
Comment (AlexanderYastrebov) [inline filters/builtin/compress.go:44]: I think `unsupported encoding` was fine

File: filters/builtin/compress.go [hunk 14]
```diff
@@ -104,47 +119,60 @@ func (e encodings) Swap(i, j int)      { e[i], e[j] = e[j], e[i] }
...
 // The compression happens in a streaming way, using only a small internal buffer.
 //
-func NewCompress() filters.Spec { return &compress{} }
+func NewCompress() filters.Spec { return NewCompressWithEncodings(supportedEncodings) }
+
+func NewCompressWithEncodings(encoding []string) filters.Spec {
+	encodingMap := map[string]int{}
+	for i, v := range encoding {
+		if !stringsContain(supportedEncodings, v) {
+			continue // skipping unsupported encodingPriority
```
Comment (AlexanderYastrebov) [inline filters/builtin/compress.go:151]: I think it is better to return error here

Comment (fpozzobon) [reply filters/builtin/compress.go]: compromising with warning log, this case would happen only if developer add a new option for compress but forget to udpate `supportedEncodings`

File: filters/builtin/compress.go (hunk 13 above)
Comment (fpozzobon) [reply filters/builtin/compress.go]: updated

File: filters/builtin/compress.go (hunk 12 above)
Comment (fpozzobon) [reply filters/builtin/compress.go]: good point, updated

File: docs/reference/filters.md (hunk 10 above)
Comment (fpozzobon) [reply docs/reference/filters.md]: added

File: filters/builtin/compress.go (hunk 5 above)
Comment (fpozzobon) [reply filters/builtin/compress.go]: it is indeed but as we need to loop through all the client supported encodings to find the "best match", the complexity is probably worth it? (implemented)

File: filters/builtin/compress.go [hunk 15]
```diff
@@ -104,47 +119,60 @@ func (e encodings) Swap(i, j int)      { e[i], e[j] = e[j], e[i] }
...
+// with chunked transfer encodingPriority, sets the Content-Encoding to the selected
+// encodingPriority and sets the Vary: Accept-Encoding header, if missing.
 //
 // The compression happens in a streaming way, using only a small internal buffer.
 //
-func NewCompress() filters.Spec { return &compress{} }
+func NewCompress() filters.Spec { return NewCompressWithEncodings(supportedEncodings) }
+
+func NewCompressWithEncodings(encoding []string) filters.Spec {
+	encodingMap := map[string]int{}
```
[the remark this answers is not in the document]
Comment (fpozzobon) [reply filters/builtin/compress.go]: replaced with m, arg is now option

File: filters/filters.go [hunk 16]
```diff
@@ -169,6 +170,9 @@ var ErrInvalidFilterParameters = errors.New("invalid filter parameters")
 
 // Registers a filter specification.
 func (r Registry) Register(s Spec) {
+	if _, ok := r[s.Name()]; ok {
+		log.Infof("Replacing registry: %s", s.Name())
```
Comment (AlexanderYastrebov) [inline filters/filters.go:174]: `name := s.Name()`
`Replacing %s filter specification`

File: filters/builtin/compress_test.go [hunk 17]
```diff
@@ -270,14 +275,14 @@ func TestCompress(t *testing.T) {
...
-		"does not accept encoding",
+		"does not accept encodingPriority",
 		http.Header{},
 		3 * 8192,
 		nil,
 		"",
 		http.Header{},
 	}, {
-		"unknown encoding",
+		"unknown encodingPriority",
```
Comment (AlexanderYastrebov) [inline filters/builtin/compress_test.go:285]: Lets revert encoding->encodingPriority changes in strings and docs: the `encodingPriority` is an internal structure name while e.g. this test verifies that unknown encoding has no effect, i.e. its not related to "priority" in any way

Comment (fpozzobon) [reply filters/builtin/compress_test.go]: good catch, bad refactoring. done

File: filters/builtin/compress.go [hunk 18]
```diff
@@ -104,47 +124,63 @@ func (e encodings) Swap(i, j int)      { e[i], e[j] = e[j], e[i] }
...
+	return NewCompressWithOptions(CompressOptions{supportedEncodings})
+}
+
+func NewCompressWithOptions(options CompressOptions) filters.Spec {
+	m := map[string]int{}
+	for i, v := range options.Encodings {
+		if !stringsContain(supportedEncodings, v) {
+			log.Warningf("Skipping unsupported encoding: %s", v)
+			continue
+		}
```
Comment (AlexanderYastrebov) [inline filters/builtin/compress.go:160]: Options are derived from user-supplied string flag, by returning a error and handling it in the startup routine we would prevent Skipper from starting with unsupported values e.g. due to typo and let users notice the problem right away instead of postponing it until they ever look into the logs.

File: filters/builtin/compress.go (hunk 5 above)
Comment (AlexanderYastrebov) [reply filters/builtin/compress.go]: The suggestion was implemented. Map lookup is cleaner and does not require auxiliary `index` method that iterates slice to find a value.

File: filters/builtin/compress.go (hunk 18 above)
Comment (fpozzobon) [reply filters/builtin/compress.go:160]: the options are already restricted in config.go `commaListFlag("gzip", "deflate", "br")`
=> the user will get an error if he tries to use a wrong option
=> the use case would be if a developer add a new option in `commaListFlag`

I understand your concern but given that this could occur only from developer mistake, maybe we can keep it simple here in place of introducing possible error returned from new?

Comment (AlexanderYastrebov) [reply filters/builtin/compress.go:160]: Ok, fair enough, I misinterpreted `commaListFlag("gzip", "deflate", "br")` as a default value while it is actually an allowlist.

Comment (AlexanderYastrebov) [reply filters/builtin/compress.go:160]: OTOH looking forward for new options we might not be so lucky to never have to return error and adding an error would be a breaking change for the `NewCompressWithOptions`

File: filters/builtin/compress_test.go [hunk 19]
```diff
@@ -373,6 +387,42 @@ func TestCompress(t *testing.T) {
 		http.Header{
 			"Content-Encoding": []string{"deflate"},
 			"Vary":             []string{"Accept-Encoding"}},
+	}, {
+		"brotli",
```
Comment (AlexanderYastrebov) [inline filters/builtin/compress_test.go:391]: I think we do not test custom encoding preference order.
Maybe we could add `encodings []string` to the test case struct and when it is not empty then redefine compress filter using `NewCompressWithOptions` like we do in skipper.go.

This way we can test that e.g. `-compress-encodings=br,gzip` indeed picks `br` over `gzip` when `Accept-Encoding` has the Chrome/Firefox default value of `gzip, deflate, br`.

File: filters/builtin/compress.go (hunk 18 above)
Comment (szuecs) [reply filters/builtin/compress.go:160]: We have to return `filters.Spec`, there is no error that we can return. Errors should be handled upfront before calling.

Comment (AlexanderYastrebov) [reply filters/builtin/compress.go:160]: > We have to return filters.Spec

Why? We can return `filters.Spec, error` from `NewCompressWithOptions` and let caller bail out on error. The existing `NewCompress` should continue to return just a spec as it uses a valid option value.

File: filters/builtin/compress_test.go (hunk 19 above)
Comment (fpozzobon) [reply filters/builtin/compress_test.go:391]: good catch, added unit test

File: filters/builtin/compress.go (hunk 18 above)
Comment (fpozzobon) [reply filters/builtin/compress.go:160]: even though there is 99% chance that it will never happen, ignoring error on NewCompress seems a bit of a trap for developer no? :)

Comment (szuecs) [reply filters/builtin/compress.go:160]: Hmm maybe you are right, but I think we have no case were we have 2 return values to instantiate a spec.
I wonder if I forgot why...

Comment (AlexanderYastrebov) [reply filters/builtin/compress.go:160]: >  ignoring error on NewCompress seems a bit of a trap

We can not change `NewCompress` signature to return error due to backwards compatibility but since it does not receive any argument we can make sure it passes safe defaults to `NewCompressWithOptions` that never cause an error.

Comment (fpozzobon) [reply filters/builtin/compress.go]: good point, updated following above suggestions.
NewCompress is indeed tested already in the unit test

**Comment by aryszka** [comment]
:+1:
